odl/space/fspace.py

```diff
@@ -25,51 +25,254 @@
+    >>> x = np.linspace(-5, 13, 10, dtype=float).reshape((2, 5))
+    >>> out = np.empty(5, dtype=float)
+    >>> step(x, out)  # returns out
+    array([ 0.,  0.,  0.,  1.,  1.])
+    """
+    def vect_decorator(func):
+        def _vect_wrapper(x, out, **kwargs):
+            if isinstance(x, np.ndarray):  # array
+                if x.ndim == 1:
+                    dim = 1
+                elif x.ndim == 2:
+                    dim = len(x)
+                    if dim == 1:
+                        x = x.squeeze()
+                else:
+                    raise ValueError('only 1- or 2-dimensional arrays '
+                                     'supported.')
+            else:  # meshgrid
+                dim = len(x)
+                if dim == 1:
+                    x = x[0]
+            if dim == 1:
+                if out is None:
+                    out = np.empty(x.size, dtype=dtype)
+                for i, xi in enumerate(x):
+                    out[i] = func(xi, **kwargs)
+                return out
+            else:
+                if is_valid_input_array(x, dim):
+                    if out is None:
+                        out = np.empty(x.shape[1], dtype=dtype)
+                    for i, pt in enumerate(x.T):
+                        out[i] = func(pt, **kwargs)
+                    return out
+                elif is_valid_input_meshgrid(x, dim):
+                    order = meshgrid_input_order(x)
+                    if out is None:
+                        out_shape = np.broadcast(*x).shape
+                        out = np.empty(out_shape, dtype=dtype, order=order)
+
+                    # TODO: find a better way to iterate
+                    vecs = vecs_from_meshgrid(x, order=order)
+                    for i, pt in enumerate(product(*vecs)):
+                        out.flat[i] = func(pt, **kwargs)
+                    return out
+                else:
+                    if out is None:
+                        try:
+                            return func(x)
+                        except (TypeError, ValueError, IndexError):
+                            raise TypeError('invalid vectorized input type.')
+                    else:
+                        raise TypeError('invalid vectorized input type.')
+
+        @wraps(func)
+        def vect_wrapper_no_out(x, **kwargs):
+            if 'out' in kwargs:
+                raise TypeError("{}() got an unexpected keyword 'out'."
+                                "".format(func.__name__))
+            return _vect_wrapper(x, None, **kwargs)
+
+        @wraps(func)
+        def vect_wrapper_pos_out(x, out, **kwargs):
+            if out is None:
+                raise ValueError('output parameter cannot be `None`.')
+            return _vect_wrapper(x, out, **kwargs)
+
+        @wraps(func)
+        def vect_wrapper_opt_out(x, out=None, **kwargs):
+            return _vect_wrapper(x, out, **kwargs)
+
+        outarg_ = str(outarg).lower()
```

Comment:
I think it's better to use the original input in the error message rather than the transformed one. People can get confused and think "This is not what I provided!"